testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Utils.java

```diff
@@ -4,27 +4,46 @@
 import java.util.regex.Pattern;
 
 public class Utils {
-    private static final String junit5Regex = "\\[class:([\\w.]+).*\\[method:(\\w+)\\(";
-    private static final Pattern junit5Pattern = Pattern.compile(junit5Regex);
+    private static final String junit5NestedRegex = "\\[nested-class:([\\w.]+)\\]";
+    private static final Pattern junit5NestedPattern = Pattern.compile(junit5NestedRegex);
+    private static final String junit5BasicRegex = "\\[class:([\\w.]+).*\\[method:([\\w().]+)";
+    private static final Pattern junit5BasicPattern = Pattern.compile(junit5BasicRegex);
     private static final String junit4Regex = "\\[test:(\\w+)\\(([\\w.]+)\\)";
     private static final Pattern junit4Pattern = Pattern.compile(junit4Regex);
 
     /**
      * Turn the uniqueId from identifier into fully qualified method name.
      *
+     * For JUnit 5 nested test:
+     * uniqueId: [engine:junit-jupiter]/[class:com.luojl.demo.InheritedTest]/[nested-class:NestedTest]/[method:NestedTestB()]
+     * fully qualified name: com.luojl.demo.InheritedTest$NestedTest#NestedTestB()
+     *
      * For JUnit 5:
      * uniqueId: [engine:junit-jupiter]/[class:com.luojl.demo.JUnit5DemoTest]/[method:TestC()]
-     * full qualified name: com.luojl.demo.JUnit5DemoTest#TestC
+     * full qualified name: com.luojl.demo.JUnit5DemoTest#TestC()
      *
      * For JUnit 4:
      * uniqueId: [engine:junit-vintage]/[runner:com.luojl.demo.JUnit4DemoTest]/[test:TestA4(com.luojl.demo.JUnit4DemoTest)]
      * full qualified name: com.luojl.demo.JUnit4DemoTest#TestA4
      */
     public static String toFullyQualifiedName(String identifierUniqueId) {
-        Matcher matcher = junit5Pattern.matcher(identifierUniqueId);
+        Matcher matcher = junit5BasicPattern.matcher(identifierUniqueId);
         if (matcher.find()) {
-            // found JUnit 5 pattern
-            return matcher.group(1) + "#" + matcher.group(2);
+            // found JUnit 5 basic pattern: class + method
+            StringBuilder sb = new StringBuilder();
+            sb.append(matcher.group(1));  // com.package.ClassName
+
+            Matcher nestedMatcher = junit5NestedPattern.matcher(identifierUniqueId);
+            while (nestedMatcher.find()) {
```

Comment:
So is it guaranteed that this find() will traverse the matched groups in order?

Reply:
Based on the method [description](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html#find()), even though a bit vague, I think yes. I can do some experiments later.

Reply:
Based on that description, I think yes too. Let me know once you've checked a few examples, then this PR should be good to merge for me.

Reply:
Cool, I will test tomorrow.

Reply:
Tested. The order is guaranteed. Refer to [testMatchNestedClassOrder](https://github.com/luoos/junit5-testrunner-demo/blob/db25b759dca7a5ec24a280766148a49fb3a8e04e/src/test/java/com/luojl/demo/UtilsTest.java#L42)